### PR TITLE
Add skeleton functions and API for refactored renderer

### PR DESCRIPTION
**Pull request by liamcervante**
This change is part of a series of changes that will create a new plan renderer that consumes the JSON structured output produced by the `-json` flag.

This change adds a basic outline of the new approach by creating structs and empty functions. 

The changes are broken up into a series of PRs that each focus on one part of the renderer to make reviewing easy. This does mean that the order matters, so you should review the PRs in the order listed below. If you prefer to view all the changes in a single (very) big PR, then skip to the last one and change it so it diffs against the main branch instead of the last PR in the chain.

The new renderer has a single entry class: `jsonformat/renderer.go`. It then makes use of two inner packages: `differ` and `change`. The `differ` package converts a single change from the provided JSON structured format into a `Change` object from the `change` package. The `change` package is then responsible for converting the `Change` objects into strings ready to be printed by the external-facing `renderer.go` functions.

This approach differs from the previous renderer where the diff calculation and the rendering happen in the same stage. Separating the diffing and rendering functionality means that going forward adding new kinds of renderers will be easier, and making changes to the format of the output plan will also be simpler.

### PR Chain

| Description         | Link                                                        | Status           |
| ------------------- | ----------------------------------------------------------- | ---------------- |
| Public JSON structs | [#32367](https://github.com/hashicorp/terraform/pull/32367) | Approved         |
| **Functions and API**   | [#32368](https://github.com/hashicorp/terraform/pull/32368) | Pending          |
| Primitives          | [#32373](https://github.com/hashicorp/terraform/pull/32373) | Pending          |
| Sensitive           | [#32375](https://github.com/hashicorp/terraform/pull/32375) | Pending          |
| Computed            | [#32378](https://github.com/hashicorp/terraform/pull/32378) | Pending          |
| Objects             | [#32391](https://github.com/hashicorp/terraform/pull/32391) | Pending          |
| Replace             | [#32392](https://github.com/hashicorp/terraform/pull/32392) | Pending          |
| Maps                | [#32397](https://github.com/hashicorp/terraform/pull/32397) | Pending          |
| Lists               | [#32401](https://github.com/hashicorp/terraform/pull/32401) | Pending          |
| Sets                | [#32409](https://github.com/hashicorp/terraform/pull/32409) | Pending          |
| Blocks              | [#32422](https://github.com/hashicorp/terraform/pull/32422) | Pending          |
| Outputs             | [#32426](https://github.com/hashicorp/terraform/pull/32426) | Pending          |

More PRs are in progress. One will port all the existing unit tests within [format/diff_test.go](https://github.com/hashicorp/terraform/blob/main/internal/command/format/diff_test.go) across to the new renderer to ensure maximum test coverage. I expect that this PR will find many missing edge cases and bugs that need to be fixed. Small bugs will be fixed as part of the main PR, but anything large will be broken out into additional PRs. Another will connect the renderer up to the rest of Terraform. This will be the last PR and it will only be merged once all possible unit tests are passing.

### CODE REVIEW

File: internal/command/jsonformat/differ/value.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+)
+
+// Value contains the unmarshalled generic interface{} types that are output by
+// the JSON structured run output functions in the various json packages (such
+// as jsonplan and jsonprovider).
+//
+// A Value can be converted into a change.Change, ready for rendering, with the
+// ComputeChangeForAttribute, ComputeChangeForOutput, and ComputeChangeForBlock
+// functions.
+type Value struct {
```
Comment (alisdair) [inline internal/command/jsonformat/differ/value.go:16]: I'd like to bikeshed about the naming here a bit, because `Value` is very generic and seems a bit odd when it actually mostly represents a before-after pair.

In another plan rendering implementation, I used the (similarly vague) term `DiffItem`. That isn't any better, so let's not do that.

I'd suggest `Change` here, but you're already using that to represent what I think is a processed, renderable diff. Maybe that type and package should be considered for a rename too…?

Wooly, tentative suggestions:

- Consider renaming `Change` to something which indicates that it is a logical difference between current and planned state. For example, `Diff` or `Plan` or some awful compound involving nothingwords like `Item`, `Container`, or `Value`.
- Alternatively, since the `change` package is where renderers live, just take the shortcut and rename it to something about rendering or views.
- Consider renaming `Value` to something which indicates that it is an unprocessed pair of before and after values from current and planned state. For example, `Change`, `Delta`, etc.
- Before doing either of the above, if you haven't already, look for existing examples of similar concepts within the codebase which closely map to these concepts, and consider reusing those names.

(These suggestions of course do not block this PR merge.)

File: internal/command/jsonformat/README.md [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+the line ending commas.
+
+##### Implementing new change types
+
+To implement a new change type, you must implement the internal Renderer 
+functionality. To do this you create a new implementation of the 
+`change/renderer.go`, make sure it accepts all the data you need, and implement
+the `Render` function (and any other additional render functions that may 
+exist). Some changes publish warnings that should be displayed alongside them, 
+if your new change has no warnings you can use the `NoWarningsRenderer` to avoid
```
Comment (alisdair) [inline internal/command/jsonformat/README.md:197]:
```suggestion
exist).

Some changes publish warnings that should be displayed alongside them. 
If your new change has no warnings you can use the `NoWarningsRenderer` to avoid
```

**Review by alisdair — APPROVED** [review]

File: internal/command/jsonformat/differ/value.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
+package differ
+
+import (
+	"encoding/json"
+
+	"github.com/hashicorp/terraform/internal/command/jsonplan"
+)
+
+// Value contains the unmarshalled generic interface{} types that are output by
+// the JSON structured run output functions in the various json packages (such
```
[the remark this answers is not in the document]
Comment (liamcervante) [reply internal/command/jsonformat/differ/value.go:10]: Cool, I've removed the structured run out bit and just said "JSON functions"

File: internal/command/jsonformat/differ/value.go (hunk 1 above)
Comment (liamcervante) [reply internal/command/jsonformat/differ/value.go:16]: Thanks for the suggestions, I've made a not of this and I will come back later and do a refactoring of the struct and package names 👍

File: internal/command/jsonformat/differ/value.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+// as jsonplan and jsonprovider).
+//
```
[the remark this answers is not in the document]
Comment (liamcervante) [reply internal/command/jsonformat/differ/value.go:12]: Done!
